### PR TITLE
ERMS-2777:

### PR DESCRIPTION
**Pull request by Moetez16**
CopyElementsIntoSubscription umbauen, damit es für andere Objekte auch nutzbar ist

### CODE REVIEW

File: app/grails-app/controllers/com/k_int/kbplus/AjaxController.groovy [hunk 1]
```diff
@@ -2678,15 +2678,21 @@ class AjaxController {
...
+        if (showActiveSubs) { queryParams.status << RDStore.SUBSCRIPTION_CURRENT.id }
+        if (showIntendedSubs) { queryParams.status << RDStore.SUBSCRIPTION_INTENDED.id }
+
         queryParams.showSubscriber = showSubscriber
         queryParams.showConnectedSubs = showConnectedSubs
 
         data = subscriptionService.getMySubscriptions_writeRights(queryParams)
 
+
+        println(data)
```
Comment (croqueGrec09) [inline app/grails-app/controllers/com/k_int/kbplus/AjaxController.groovy:2695]: debug residual?

Comment (Moetez16) [reply app/grails-app/controllers/com/k_int/kbplus/AjaxController.groovy:2695]: Upps

File: app/grails-app/views/templates/copyElements/_copyElements.gsp [hunk 2]
```diff
@@ -1,28 +1,29 @@
...
             method="post" class="ui form newLicence">
+        <input type="hidden" name="${FormService.FORM_SERVICE_TOKEN}" value="${formService.getNewToken()}"/>
         <table class="ui celled table table-tworow la-table">
             <thead>
             <tr>
                 <th class="six wide">
-                    <g:if test="${sourceSubscription}"><g:link controller="subscription" action="show"
-                                                               id="${sourceSubscription.id}">${sourceSubscription.dropdownNamingConvention()}</g:link></g:if>
+                    <g:if test="${sourceObject}"><g:link controller="subscription" action="show"
+                                                               id="${sourceObject.id}">${sourceObject.dropdownNamingConvention()}</g:link></g:if>
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/copyElements/_copyElements.gsp:26]: Beware: not every object does have a dropdownNamingConvention()

File: app/grails-app/views/templates/subscription/_copyDocsAndTasks.gsp [hunk 3]
```diff
@@ -4,31 +4,31 @@
...
                     <th class="six wide">
-                        <g:if test="${sourceSubscription}"><g:link controller="subscription" action="show" id="${sourceSubscription?.id}">${sourceSubscription?.dropdownNamingConvention()}</g:link></g:if>
+                        <g:if test="${sourceObject}"><g:link controller="subscription" action="show" id="${sourceObject?.id}">${sourceObject?.dropdownNamingConvention()}</g:link></g:if>
                     </th>
                     <th class="one wide center aligned">
                         <input type="checkbox" name="checkAllCopyCheckboxes" data-action="copy" onClick="toggleAllCheckboxes(this)" checked />
                     </th>
                     <th class="six wide">
-                        <g:if test="${targetSubscription}"><g:link controller="subscription" action="show" id="${targetSubscription?.id}">${targetSubscription?.dropdownNamingConvention()}</g:link></g:if>
+                        <g:if test="${targetObject}"><g:link controller="subscription" action="show" id="${targetObject?.id}">${targetObject?.dropdownNamingConvention()}</g:link></g:if>
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copyDocsAndTasks.gsp:28]: see above

File: app/grails-app/views/templates/copyElements/_selectSourceAndTargetObject.gsp [hunk 4]
```diff
@@ -1,44 +1,50 @@
...
                     <label for="show.conntectedSubscriptions">${message(code:'subscription.details.copyElementsIntoSubscription.show.conntectedSubscriptions.name')}</label>
                 </div><br id="element-vor-target-dropdown" />
                 <g:select class="ui search selection dropdown"
-                      id="targetSubscriptionId"
-                      name="targetSubscriptionId"
-                      from="${allSubscriptions_writeRights}"
+                      id="targetObjectId"
+                      name="targetObjectId"
+                      from="${allObjects_writeRights}"
                       optionValue="${{it?.dropdownNamingConvention()}}"
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/copyElements/_selectSourceAndTargetObject.gsp:45]: it?, then: it does not need to have a dropdownNamingConvention()

File: app/grails-app/views/templates/subscription/_copyElements.gsp [hunk 5]
```diff
@@ -550,10 +550,10 @@
                         </b>
                         <g:each in="${target_visibleOrgRelations}" var="target_role">
                             <g:if test="${target_role.org}">
-                                <g:if test="${Person.getPrivateByOrgAndObjectRespFromAddressbook(target_role.org, targetSubscription, 'Specific subscription editor', contextService.getOrg()) ||
-                                        Person.getPublicByOrgAndObjectResp(target_role.org, targetSubscription, 'Specific subscription editor')}">
+                                <g:if test="${Person.getPrivateByOrgAndObjectRespFromAddressbook(target_role.org, targetObject, 'Specific subscription editor', contextService.getOrg()) ||
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copyElements.gsp:553]: Specific subscription editor for objects other than subscription?

File: app/grails-app/views/templates/subscription/_copyPackagesAndIEs.gsp [hunk 6]
```diff
@@ -4,23 +4,23 @@
...
+            params="[workFlowPart: workFlowPart, sourceObjectId: sourceObjectId, targetObjectId: targetObject?.id, isRenewSub: isRenewSub, fromSurvey: fromSurvey]"
             method="post" class="ui form newLicence">
         <table class="ui celled table table-tworow la-table">
             <thead>
                 <tr>
                     <th class="six wide">
                         <div class="la-copyElements-th-flex-container">
                             <div class="la-copyElements-th-flex-item">
-                                <g:if test="${sourceSubscription}"><g:link controller="subscription" action="show" id="${sourceSubscription?.id}">${sourceSubscription?.dropdownNamingConvention()}</g:link></g:if>
+                                <g:if test="${sourceObject}"><g:link controller="subscription" action="show" id="${sourceObject?.id}">${sourceObject?.dropdownNamingConvention()}</g:link></g:if>
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copyPackagesAndIEs.gsp:23]: again ... sourceObject does not necessarily have a dropdownNamingConvention()

File: app/grails-app/views/templates/subscription/_copyPackagesAndIEs.gsp [hunk 7]
```diff
@@ -62,7 +62,7 @@
                                                 <g:message code="subscription.packages.${pcc.settingKey}"/>: ${pcc.settingValue ? pcc.settingValue.getI10n('value') : RDStore.PENDING_CHANGE_CONFIG_PROMPT.getI10n('value')} (<g:message code="subscription.packages.notification.label"/>: ${pcc.withNotification ? RDStore.YN_YES.getI10n('value') : RDStore.YN_NO.getI10n('value')})
                                                 <g:if test="${accessService.checkPermAffiliation('ORG_CONSORTIUM','INST_EDITOR')}">
                                                     <g:if test="${!(pcc.settingKey in excludes)}">
-                                                        <g:if test="${auditService.getAuditConfig(sourceSubscription,pcc.settingKey)}">
+                                                        <g:if test="${auditService.getAuditConfig(sourceObject,pcc.settingKey)}">
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copyPackagesAndIEs.gsp:65]: Only subscriptions (precisely SubscriptionPackages) may have pendingChangeConfigurations

File: app/grails-app/views/templates/subscription/_copyPropertiesCompare.gsp [hunk 8]
```diff
@@ -9,20 +9,20 @@
     <body>
     <semui:form>
         <g:if test="${!fromSurvey && !isRenewSub}">
-            <g:render template="selectSourceAndTargetSubscription" model="[
-                sourceSubscription: sourceSubscription,
-                targetSubscription: targetSubscription,
-                allSubscriptions_readRights: allSubscriptions_readRights,
-                allSubscriptions_writeRights: allSubscriptions_writeRights]"/>
+            <g:render template="/templates/copyElements/selectSourceAndTargetObject" model="[
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copyPropertiesCompare.gsp:12]: Apart from renaming, this view has not been generalised at all

File: app/grails-app/views/templates/subscription/_copySubscriber.gsp [hunk 9]
```diff
@@ -26,9 +26,9 @@
                                 <thead>
                                 <tr>
                                     <th colspan="5">
-            <g:if test="${sourceSubscription}"><g:link controller="subscription"
+            <g:if test="${sourceObject}"><g:link controller="subscription"
                                                        action="show"
-                                                       id="${sourceSubscription?.id}">${sourceSubscription?.dropdownNamingConvention()}</g:link></g:if>
+                                                       id="${sourceObject?.id}">${sourceObject?.dropdownNamingConvention()}</g:link></g:if>
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_copySubscriber.gsp:31]: sourceObject does not necessarily have a dropdownNamingConvention

File: app/grails-app/views/templates/subscription/_propertyComparisonTableRow.gsp [hunk 10]
```diff
@@ -60,11 +60,11 @@
             </g:if>
         </td>
         <g:set var="propValues" value="${prop.getValue()}" />
-        <% Set propValuesForSourceSub = propValues.get(sourceSubscription) %>
-        <% Set propValuesForTargetSub = propValues.get(targetSubscription) %>
+        <% Set propValuesForSourceSub = propValues.get(sourceObject) %>
+        <% Set propValuesForTargetSub = propValues.get(targetObject) %>
         %{--SOURCE-SUBSCRIPTION--}%
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/subscription/_propertyComparisonTableRow.gsp:65]: Should it not be extended to SOURCE-OBJECT?

**Pull request merged**